plugin/pkg/replacer/replacer.go

```diff
@@ -41,168 +56,247 @@ var labels = []string{
-func addrToRFC3986(addr string) string {
-	if strings.Contains(addr, ":") {
-		return "[" + addr + "]"
+// appendAddrToRFC3986 will add brackets to the address if it is an IPv6 address.
+func appendAddrToRFC3986(b []byte, addr string) []byte {
+	if strings.IndexByte(addr, ':') != -1 {
+		b = append(b, '[')
+		b = append(b, addr...)
+		b = append(b, ']')
+	} else {
+		b = append(b, addr...)
 	}
-	return addr
+	return b
 }
 
+type nodeType int
+
 const (
-	headerReplacer = "{>"
-	labelReplacer  = "{/"
-	// EmptyValue is the default empty value.
-	EmptyValue = "-"
+	typeLabel    nodeType = iota // "{type}"
+	typeLiteral                  // "foo"
+	typeMetadata                 // "{/metadata}"
 )
+
+// A node represents a segment of a parsed format.  For example: "A {type}"
+// contains two nodes: "A " (literal); and "{type}" (label).
+type node struct {
+	value string // Literal value, label or metadata label
+	typ   nodeType
+}
+
+// A replacer is an ordered list of all the nodes in a format.
+type replacer []node
+
+func parseFormat(s string) replacer {
+	// Assume there is a literal between each label - its cheaper to over
+	// allocate once than allocate twice.
+	rep := make(replacer, 0, strings.Count(s, "{")*2)
+	for {
+		// We find the right bracket then backtrack to find the left bracket.
+		// This allows us to handle formats like: "{ {foo} }".
+		j := strings.IndexByte(s, '}')
+		if j < 0 {
+			break
+		}
+		i := strings.LastIndexByte(s[:j], '{')
+		if i < 0 {
+			// Handle: "A } {foo}"
+			//            ^j
+			//
+			// By treating "A }" as a literal
+			rep = append(rep, node{
+				value: s[:j+1],
+				typ:   typeLiteral,
+			})
+			s = s[j+1:]
+			continue
+		}
+
+		val := s[i : j+1]
+		var typ nodeType
+		switch _, ok := knownLabels[val]; {
+		case ok:
+			typ = typeLabel
+		case strings.HasPrefix(val, "{/"):
+			// Strip "{/}" from metadata labels
+			val = val[2 : len(val)-1]
+			typ = typeMetadata
+		default:
+			// Given: "A {X}" val is "{X}" expand it to the whole literal.
+			val = s[:j+1]
+			typ = typeLiteral
+		}
+
+		// Append any leading literal.  Given "A {type}" the literal is "A "
+		if i != 0 && typ != typeLiteral {
+			rep = append(rep, node{
+				value: s[:i],
+				typ:   typeLiteral,
+			})
+		}
+		rep = append(rep, node{
+			value: val,
+			typ:   typ,
+		})
+		s = s[j+1:]
+	}
+	if len(s) != 0 {
+		rep = append(rep, node{
+			value: s,
+			typ:   typeLiteral,
+		})
+	}
+	return rep
+}
+
+var replacerCache sync.Map // map[string]replacer
+
+func loadFormat(s string) replacer {
+	if v, ok := replacerCache.Load(s); ok {
+		return v.(replacer)
+	}
+	v, _ := replacerCache.LoadOrStore(s, parseFormat(s))
+	return v.(replacer)
+}
+
+// bufPool stores pointers to scratch buffers
+var bufPool = sync.Pool{
+	New: func() interface{} {
+		b := make([]byte, 0, 256)
+		return &b
+	},
+}
+
+func (r replacer) Replace(ctx context.Context, state request.Request, rr *dnstest.Recorder) string {
+	p := bufPool.Get().(*[]byte)
```

Comment:
you have a pointer to a slice, so you can extend the slice or something? This requires some documentation detailing why it's a pointer to a slice. Probably where the syncmap is declared.

Reply:
I'll add a comment.  The reason we use store a pointer to the slice and not the slice itself is that it saves an alloc in the value => `interface{}` conversion.

Reply:
Comment added to bufPool https://github.com/coredns/coredns/blob/57fa92527aa00dc0b38db10eda5fd204218f1d46/plugin/pkg/replacer/replacer.go#L270-L274

Reply:
> [ Quoting <notifications@github.com> in "Re: [coredns/coredns] replacer: eva..." ]
> I'll add a comment. The reason we use store a pointer to the slice and not the slice itself is that it saves an alloc in the value => `interface{}` conversion.
> this sounds like a really micro optimizations we shouldn't really care about. I doubt this makes a dent in the 3x improvement we're seeing here. Logging will always kill your performance, doing these micro optimizations won't make it that much better. Just use the slice directly. /Miek
> […](#)
> -- Miek Gieben

IMO, more a matter of correctness, but I'll change it to use the value so we don't have to deal with pointers.